### ISSUE TITLE
URL containing %2B incorrectly handled in RAW proxy

### ISSUE THREAD

**Issue by filo891** [sonatype/nexus-public#385]
* What problem are you trying to solve?
I have configured a RAW proxy pointing to https://download.pytorch.org/whl/. Some files at the remote destination contain "%2B" in their names, which gets automatically decoded by Nexus to "+" when proxying, resulting in the file not being found at the remote location.

Example:
```
# Downloading the existing file works
> curl -I https://download.pytorch.org/whl/rocm4.2/torch-1.10.2%2Brocm4.2-cp36-cp36m-linux_x86_64.whl
HTTP/2 200

# Downloading the file with a '+' instead of '%2B' in the name does not work
> curl -I https://download.pytorch.org/whl/rocm4.2/torch-1.10.2+rocm4.2-cp36-cp36m-linux_x86_64.whl
HTTP/2 403

# Downloading the file through the Nexus RAW proxy does not work
> date && curl -I http://localhost:8080/nexus/repository/pytorch/whl/rocm4.2/torch-1.10.2%2Brocm4.2-cp36-cp36m-linux_x86_64.whl
Wed Apr 24 18:05:03 CEST 2024
HTTP/1.1 404 Not Found
Server: Nexus/3.67.1-01 (OSS)

# In the outbound-request.log it appears Nexus translated "%2B" to "+":
[24/Apr/2024:18:05:03 +0200] - "GET https://download.pytorch.org/whl/rocm4.2/torch-1.10.2+rocm4.2-cp36-cp36m-linux_x86_64.whl HTTP/1.1" 403 -1 486 "Nexus/3.67.1-01 (OSS; Linux; 3.10.0-1160.114.2.el7.x86_64; amd64; 1.8.0_402)" [qtp1346273897-19716]
```

* Do you have a workaround you are using at present?
No

* What feature or behavior is this required for?
RAW repository proxy

* How could we solve this issue? (Not knowing is okay!)
Do not automatically translate (perhaps make it as a selectable option) escaped URLs when making the backend request.

* Tell us about your Nexus Repository deployment: what version, operating system, and database are you using?
OSS 3.67.1-01 on Linux

**Comment by nblair** [comment]
Hi @filo891 - thanks for opening an issue. We are aware of a limited issues where paths including URL characters that require careful encoding don't work with a Raw Proxy. 

I'm unfamiliar with the download site you are referencing, but it appears to provide python packages. Instead of using a raw proxy, is it possible to configure a pypi proxy for that distributor's content instead as a workaround?

**Comment by filo891** [comment]
Hi @nblair,

Initially I tried with a PyPi proxy, but it was not working as expected, so I have opted for the RAW proxy.

Perhaps I was doing something wrong and you can help me out here. Packages from teh source repository are installed e.g. using the following command:
```pip3 install torch --index-url https://download.pytorch.org/whl```

I have configured a PyPi proxy (named "pytorch-test") in Nexus and pointed it to [https://download.pytorch.org/](https://download.pytorch.org/). then tried to install the package through the proxy:
```
pip3 install torch --index-url https://nexus.example.com:8080/nexus/repository/pytorch-test/whl
Collecting torch
  Could not find a version that satisfies the requirement torch (from versions: )
No matching distribution found for torch
```

Do you know if there is something I should be doing differently?
Thank you.

**Comment by nblair** [comment]
I tried to learn a little about this pytorch site, and a few things came up. https://download.pytorch.org/ is not a pypi registry of any sort, it's just a website with some direct downloads.

Now [pytorch is actually published at pypi.org](https://pypi.org/project/torch/), the official pypi registry. [Pytorch's getting started instructions](https://pytorch.org/get-started/locally/#mac-anaconda) show it can be installed with a simple `pip3 install torch`, which by default will instruct the pypi client to download it from pypi.org.

In order to proxy that request through Nexus Repository, create a pypi repository proxying https://pypi.org/, then follow the [client instructions](https://help.sonatype.com/en/pypi-repositories.html) we have to configure pip accordingly. That should do the trick.

**Comment by filo891** [comment]
Not all versions of the packages are available on pypi.org. The page you linked suggests installing them from https://download.pytorch.org/ for some combinations of OS/Compute Platform (see: https://pytorch.org/get-started/locally/#start-locally).

So I'm afraid the RAW proxy repository is my only option to get the packages into Nexus. Is there a chance to get the URL encoding fixed? I'm happy to contribute with a pull request if you point me to the relevant part of the code.

**Comment by filo891** [comment]
Hi @nblair,

Would it be possible to provide an update on this? As I mentioned I'm happy to contribute fixing the issue, just a few pointers would be needed so spare my time.

Thank you.

**Comment by teamsmiley** [comment]
i have same issue  i proxy to oracle on my nexus 

```
curl -I https://yum.oracle.com/repo/OracleLinux/OL8/baseos/latest/x86_64/getPackage/perl-Text-Tabs+Wrap-2013.0523-395.el8.noarch.rpm
```

filename has '+'. it is not working. 

any work around?

**Comment by DougTea** [comment]
same issue,need help

**Comment by lbernard-fr** [comment]
Same issue

**Comment by siameseoriental** [comment]
Hi to all!
Probably it would be helpful to somebody
As a workaround I used Nginx to replace `+` with `%2B`
Here is working config:
```
http {
    resolver 8.8.8.8 ipv6=off valid=300s;
    map $request_uri $final_uri {
         "~^(.*)\+(.*)$"  "$1%2B$2";
         default          $request_uri;
    }

    server {
         listen 80;
         server_name pytorch.mysite.local;
         location /whl/ {
             proxy_pass https://download.pytorch.org$final_uri;
             proxy_set_header Host download.pytorch.org;
             proxy_ssl_server_name on;
             proxy_ssl_name download.pytorch.org;
         }
    }
}
```
In the Nexus repository I used http://pytorch.mysite.local as a target

**Comment by benjaminwilcox** [comment]
Same issue, any updates, @nblair?

**Comment by eshack94** [comment]
> Hi @filo891 - thanks for opening an issue. We are aware of a limited issues where paths including URL characters that require careful encoding don't work with a Raw Proxy.
> 
> I'm unfamiliar with the download site you are referencing, but it appears to provide python packages. Instead of using a raw proxy, is it possible to configure a pypi proxy for that distributor's content instead as a workaround?

@nblair No, it is not. Because these sites do not follow PEP-503, meaning they don't support the `/simple` endpoint, etc. And Nexus doesn't support Pypi repos that aren't compliant with this standard. See: https://help.sonatype.com/en/pypi-repositories.html


From the docs:

```
Note

Nexus Repository currently requires the remote PyPI repository to support the /simple endpoint.

Repositories that do not follow this convention, such as https://pypi.nvidia.com, are not supported at this time.

```

**Comment by alexbevz** [comment]
same problem :(

**Comment by crbetts1** [comment]
Any update on this, having problems with yum repo and gcc-c++ package

**Comment by leodevian** [comment]
Check your reverse proxies. Network administrators will block encoded characters in URLs for security reasons.

**Comment by crbetts1** [comment]
@leodevian The outbound request logged in Nexus is converting what yum has sent back to a '+' sign as an example:

[01/Apr/2026:15:52:41 +0000] - "GET https://al2023-repos-us-east-1.s3.dualstack.us-east-1.amazonaws.com/blobstore/b7aeb68c966b50784c0e0d3dfedf4d6c41/gcc-c++-11.5.0-5.amzn2023.0.5.x86_64.rpm HTTP/1.1" 403 -1 14 --- [qtp1416321340-162]
